docs/enhancements/auto-return-agent-to-infra-env.md

```diff
@@ -0,0 +1,200 @@
+6. Discovery agent registers host
+  * Service moves host to `HostStatusDiscoveringUnbound`
+
+### User Stories
+
+#### Story 1
+
+As a cluster creator I want to use the assisted service to repurpose hosts
+from existing clusters without requiring access to the host management
+infrastructure (BMC, hypervisor, etc).
+
+#### Story 2
+
+As a cluster creator, I want the host's storage wiped when it is removed from a
+cluster to avoid data leaking to other people who create and use clusters in the
+same environment.
+
+### Implementation Details/Notes/Constraints [optional]
+
+* The boot artifact URLs (initrd, kernel, and rootfs) will be provided to the
+  agent directly as a part of the machine config that runs the agent.
+* The agent will require one new command which will be executed when a host is
+  in `HostStatusReclaiming`
+  * It will handle steps 3, 4, and 5 from the "Timeline" section
+* The exiting unbind API will be used to trigger this process
+  * This means removing the cluster ref from the agent CR when using the kubeAPI
+    and the existing unbind API in the REST API.
+
+#### Example boot entry
+
+```
+$ cat /boot/loader/entries/00-assisted-discovery.conf
+title Assisted Installer Discovery
+version 999
+options random.trust_cpu=on ignition.firstboot ignition.platform.id=metal 'coreos.live.rootfs_url=http://10.1.155.24:6016/boot-artifacts/rootfs?arch=x86_64&version=4.10'
+linux /boot/discovery/vmlinuz
+initrd /boot/discovery/c4bd9751-f3ae-49ac-a36b-d3ffebef4a39-initrd.img
+```
+
+#### State Machine Changes
+Currently when a host is installed, all unbind APIs will call the `UnbindHost`
+transition. This will always move the host to
+`HostStatusUnbindingPendingUserAction`. If the agent has a BMH associated with
+it, the bare metal host agent controller (BMAC) will make the changes required
+to boot the host back into discovery. If there is no BMH associated with the
+agent (or the kube API is not in use), the host will remain in
+`HostStatusUnbindingPendingUserAction`.
+
+The following diagram illustrates the new proposed transition to indicate that a
+running agent should execute the reclaim procedure.
+
+```
+                   -----------------------
+                   | HostStatusInstalled |
+                   -----------------------
+     REST API or no BMH      |   Kube API with BMH
+          -------------------+-------------------
+          |                                     |
+          V                                     V
+-----------------------------            ----------------------------
+| TransitionTypeReclaimHost |            | TransitionTypeUnbindHost |
+-----------------------------            ----------------------------
+            |                                               |
+            V                                               |
+  ------------------------                                  |
+  | HostStatusReclaiming |                                  |
```

Comment:
Do we do anything when the agent registers here? Or just no-op that transition?

Reply:
@filanov?

Reply:
I think it's probably best to add another status after this one.

We'll move to `HostStatusReclaiming` when a user unbinds. The agent will be given a step to set up the host for booting back into discovery.
Move from `HostStatusReclaiming` to `HostStatusReclaimReboot` when the agent reports that the artifacts and boot entry have been set up properly (posted step response)
Move from `HostStatusReclaimReboot` to `HostStatusDiscoveringUnbound` when the host registers.

If the host fails in Reclaiming or times out in ReclaimReboot we'll move to `HostStatusUnbindingPendingUserAction`

This means we can avoid creating a reclaim-specific status column in the host and will more clearly indicate that the host is rebooting.
This also reserves UnbindingPendingUserAction for cases where user (or bmac) action is actually required. We can also continue to use this state for our error case.

What do you think @filanov ?

Reply:
Added another state below to track reboot.